// mozaik-ext-aws/lib/components/Instances.js
"use strict";

var React = require("react");
var Reflux = require("reflux");
var _ = require("lodash");
var ApiConsumerMixin = require("mozaik/browser").Mixin.ApiConsumer;

var Instances = React.createClass({
    displayName: "Instances",
    mixins: [Reflux.ListenerMixin, ApiConsumerMixin],

    getInitialState: function getInitialState() {
        return {
            instances: []
        };
    },

    getApiRequest: function getApiRequest() {
        return {
            id: "aws.instances"
        };
    },

    onApiData: function onApiData(instances) {
        var _this = this;
        // if we have an available filter on instance name, apply it
        if (this.props.nameFilter) {
            instances = _.where(instances, function (instance) {
                return _this.props.nameFilter.test(instance.name);
            });
        }

        this.setState({
            instances: instances
        });
    },

    render: function render() {
        var instanceNodes = _.map(this.state.instances, function (instance) {
            var cssClass = "aws__instance aws__instance--" + instance.state;

            return React.createElement(
                "div",
                { key: instance.id, className: cssClass },
                instance.name,
                instance.state,
                React.createElement(
                    "span",
                    { className: "aws__instance__id" },
                    instance.id
                )
            );
        });

        return React.createElement(
            "div",
            null,
            React.createElement(
                "div",
                { className: "widget__header" },
                "AWS instances",
                React.createElement(
                    "span",
                    { className: "widget__header__count" },
                    this.state.instances.length
                ),
                React.createElement("i", { className: "fa fa-hdd-o" })
            ),
            React.createElement(
                "div",
                { className: "widget__body" },
                instanceNodes
            )
        );
    }
});

module.exports = Instances;
