// mozaik-ext-aws/lib/components/Stacks.js
"use strict";

var React = require("react");
var Reflux = require("reflux");
var _ = require("lodash");
var Stack = require("./Stack");
var ApiConsumerMixin = require("mozaik/browser").Mixin.ApiConsumer;

var Stacks = React.createClass({
    displayName: "Stacks",
    mixins: [Reflux.ListenerMixin, ApiConsumerMixin],

    getInitialState: function getInitialState() {
        return {
            stacks: []
        };
    },

    getApiRequest: function getApiRequest() {
        return {
            id: "aws.stacks"
        };
    },

    onApiData: function onApiData(stacks) {
        this.setState({
            stacks: stacks
        });
    },

    render: function render() {
        var stackNodes = _.map(this.state.stacks, function (stack) {
            return React.createElement(Stack, { key: stack.StackId, stack: stack });
        });

        return React.createElement(
            "div",
            null,
            React.createElement(
                "div",
                { className: "widget__header" },
                "AWS stacks",
                React.createElement(
                    "span",
                    { className: "widget__header__count" },
                    this.state.stacks.length
                ),
                React.createElement("i", { className: "fa fa-cloud" })
            ),
            React.createElement(
                "div",
                { className: "widget__body" },
                stackNodes
            )
        );
    }
});

module.exports = Stacks;
